From 44de47fba8e696ee5f6bdc074070805cddb99d8c Mon Sep 17 00:00:00 2001
From: Tim Nordell <tnordell@airgain.com>
Date: Wed, 11 Mar 2026 22:18:11 -0500
Subject: [PATCH 4/7] jail ubus: Add in ubus object ID cache

In order to do ACLs, we need to be able to look up UBUS objects by ID
to find out their name.  This API does not exist in the ubus daemon, and
it'll be useful to not have to rely on calling out to ubus to determine
these names.
---
 CMakeLists.txt           |   2 +-
 jail/ubus.c              |  62 +++++++++++++
 jail/ubus_obj_id_cache.c | 185 +++++++++++++++++++++++++++++++++++++++
 jail/ubus_obj_id_cache.h |  41 +++++++++
 4 files changed, 289 insertions(+), 1 deletion(-)
 create mode 100644 jail/ubus_obj_id_cache.c
 create mode 100644 jail/ubus_obj_id_cache.h

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1934170..f208035 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -108,7 +108,7 @@ SET(SOURCES_OCI_SECCOMP jail/seccomp-oci.c)
 ENDIF()
 
 IF(JAIL_SUPPORT)
-ADD_EXECUTABLE(ujail jail/jail.c jail/cgroups.c jail/cgroups-bpf.c jail/elf.c jail/fs.c jail/capabilities.c jail/netifd.c jail/ubus.c ${SOURCES_OCI_SECCOMP})
+ADD_EXECUTABLE(ujail jail/jail.c jail/cgroups.c jail/cgroups-bpf.c jail/elf.c jail/fs.c jail/capabilities.c jail/netifd.c jail/ubus.c jail/ubus_obj_id_cache.c ${SOURCES_OCI_SECCOMP})
 TARGET_LINK_LIBRARIES(ujail ${ubox} ${ubus} ${uci} ${blobmsg_json})
 INSTALL(TARGETS ujail
 	RUNTIME DESTINATION ${CMAKE_INSTALL_SBINDIR}
diff --git a/jail/ubus.c b/jail/ubus.c
index 687b0dc..cafa558 100644
--- a/jail/ubus.c
+++ b/jail/ubus.c
@@ -28,12 +28,17 @@
 #include <ubusmsg.h>
 
 #include "ubus.h"
+#include "ubus_obj_id_cache.h"
 #include "log.h"
 
 static const char UBUS_UNIX_SOCKET[] = "/var/run/ubus/ubus.sock";
 
 struct ubus_relay_server {
+	struct ubus_context ubus;
+	struct uloop_timeout ubus_reconnect_tmo;
 	struct uloop_fd fd;
+	struct ubus_obj_id_cache id_cache;
+	uint32_t session_obj_id;
 };
 
 struct ubus_forward_fd_ctx
@@ -376,6 +381,54 @@ static void ubus_relay_server_cb(struct uloop_fd *fd, unsigned int events)
 	return;
 }
 
+static void ubus_relay_ubus_disconnected(struct ubus_context *ctx) {
+	struct ubus_relay_server *server = container_of(ctx, struct ubus_relay_server, ubus);
+
+	INFO("Disconnected from ubus\n");
+	uloop_fd_delete(&ctx->sock);
+	uloop_timeout_set(&server->ubus_reconnect_tmo, 1000);
+	ubus_obj_id_cache_ubus_disconnected(&server->id_cache, &server->ubus);
+}
+
+static void ubus_relay_ubus_connected(struct ubus_relay_server *server) {
+	INFO("Reconnected to ubus\n");
+	server->session_obj_id = 0;
+	uloop_fd_add(&server->ubus.sock, ULOOP_BLOCKING | ULOOP_READ | ULOOP_ERROR_CB);
+	ubus_obj_id_cache_ubus_connected(&server->id_cache, &server->ubus);
+	server->ubus.connection_lost = ubus_relay_ubus_disconnected;
+}
+
+static void ubus_relay_ubus_timeout(struct uloop_timeout *ctx) {
+	struct ubus_relay_server *server = container_of(ctx, struct ubus_relay_server, ubus_reconnect_tmo);
+	if (ubus_connect_ctx(&server->ubus, NULL) == 0) {
+		ubus_relay_ubus_connected(server);
+	} else {
+		uloop_timeout_set(&server->ubus_reconnect_tmo, 1000);
+	}
+}
+
+static void ubus_relay_monitor_objs(
+	struct ubus_obj_id_cache *cache,
+	enum UBUS_OBJ_ID_CACHE method,
+	uint32_t id,
+	const char *path
+) {
+	struct ubus_relay_server *server = container_of(cache, struct ubus_relay_server, id_cache);
+	if (strcmp(path, "session") == 0) {
+		switch(method)
+		{
+			case UBUS_OBJ_ID_CACHE_ADDED:
+				INFO("Session Object ID: %08x\n", id);
+				server->session_obj_id = id;
+				break;
+			case UBUS_OBJ_ID_CACHE_REMOVED:
+				INFO("Session Object is gone\n");
+				server->session_obj_id = 0;
+				break;
+		}
+	}
+}
+
 bool ubus_create_relay(pid_t pid)
 {
 	struct ubus_relay_server *server;
@@ -411,5 +464,14 @@ bool ubus_create_relay(pid_t pid)
 		return false;
 	}
 
+	ubus_obj_id_cache_init(&server->id_cache, ubus_relay_monitor_objs);
+
+	/* Note: We'd use ubus_auto_connect(...) but it's currently broken
+	* since the uloop_fd_add() call added doesn't pass in ULOOP_ERROR_CB.
+	*/
+	server->ubus.connection_lost = ubus_relay_ubus_disconnected;
+	server->ubus_reconnect_tmo.cb = ubus_relay_ubus_timeout;
+	uloop_timeout_set(&server->ubus_reconnect_tmo, 0);
+
 	return true;
 }
diff --git a/jail/ubus_obj_id_cache.c b/jail/ubus_obj_id_cache.c
new file mode 100644
index 0000000..9352ec3
--- /dev/null
+++ b/jail/ubus_obj_id_cache.c
@@ -0,0 +1,185 @@
+/*
+ * Copyright (C) 2026 Tim Nordell <tnordell@airgain.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include "ubus_obj_id_cache.h"
+#include "log.h"
+
+#include <assert.h>
+#include <libubus.h>
+
+struct ubus_object_id
+{
+	struct avl_node avl;
+	char path[0];
+};
+
+static int avl_uintptr_cmp(const void *k1, const void *k2, void *ptr)
+{
+	uintptr_t left = (uintptr_t)k1;
+	uintptr_t right = (uintptr_t)k2;
+
+	if (left < right)
+		return -1;
+	else
+		return right < left;
+}
+
+static void avl_object_remove(struct ubus_obj_id_cache *server, uintptr_t id)
+{
+	struct ubus_object_id *node = avl_find_element(&server->avl, (void *)id, node, avl);
+	if (node)
+	{
+		avl_delete(&server->avl, &node->avl);
+		free(node);
+	}
+}
+
+static void avl_object_add(struct ubus_obj_id_cache *server, uintptr_t id, const char *path)
+{
+	int path_len = strlen(path) + 1; // Include final NUL
+	struct ubus_object_id *obj = calloc_a(sizeof(struct ubus_object_id) + path_len);
+
+	memcpy(obj->path, path, path_len);
+
+	// Check if node already exists
+	avl_object_remove(server, id);
+
+	obj->avl.key = (void *)(id);
+	avl_insert(&server->avl, &obj->avl);
+}
+
+static void initial_object_list_cb(struct ubus_context *ctx, struct ubus_object_data *obj, void *priv)
+{
+	struct ubus_obj_id_cache *server = priv;
+	avl_object_add(server, obj->id, obj->path);
+	if (server->cb)
+		server->cb(server, UBUS_OBJ_ID_CACHE_ADDED, obj->id, obj->path);
+}
+
+static void ubus_event_cb(struct ubus_context *ctx, struct ubus_event_handler *ev,
+						const char *type, struct blob_attr *msg)
+{
+	struct ubus_obj_id_cache *server = container_of(ev, struct ubus_obj_id_cache, ev);
+	static const struct blobmsg_policy policy[2] = {
+		{"id", BLOBMSG_TYPE_INT32},
+		{"path", BLOBMSG_TYPE_STRING},
+	};
+	struct blob_attr *attr[ARRAY_SIZE(policy)];
+
+	bool is_add = false;
+	bool is_remove = false;
+
+	is_add = (strcmp(type, "ubus.object.add") == 0);
+	if (!is_add)
+		is_remove = (strcmp(type, "ubus.object.remove") == 0);
+
+	if(!is_add && !is_remove)
+		return;
+
+	blobmsg_parse(policy, ARRAY_SIZE(policy), attr, blob_data(msg), blob_len(msg));
+	if (!attr[0] || !attr[1])
+		return;
+
+	uint32_t id = blobmsg_get_u32(attr[0]);
+	const char *path = blobmsg_get_string(attr[1]);
+
+	if(is_add) {
+		avl_object_add(server, id, path);
+		if (server->cb)
+			server->cb(server, UBUS_OBJ_ID_CACHE_ADDED, id, path);
+	} else {
+		avl_object_remove(server, id);
+		if (server->cb)
+			server->cb(server, UBUS_OBJ_ID_CACHE_REMOVED, id, path);
+	}
+}
+
+void ubus_obj_id_cache_init(
+	struct ubus_obj_id_cache *ctx,
+	ubus_obj_id_cache_cb cb
+) {
+	// We depend on pointers being at least equal to a uint32_t since we're
+	// using the pointer value inside the AVL tree as the actual value being
+	// stored instead of pointing to an object.
+	assert(sizeof(uint32_t) <= sizeof(uintptr_t));
+
+	memset(ctx, 0, sizeof(*ctx));
+	ctx->ubus = NULL;
+	ctx->cb = cb;
+	avl_init(&ctx->avl, avl_uintptr_cmp, false, NULL);
+}
+
+void ubus_obj_id_cache_ubus_connected(struct ubus_obj_id_cache *ctx, struct ubus_context *ubus)
+{
+	struct ubus_object_id *node, *tmp;
+	ctx->ubus = ubus;
+
+	/* Reset cache */
+	avl_remove_all_elements(&ctx->avl, node, avl, tmp) {
+		if (ctx->cb)
+			ctx->cb(ctx, UBUS_OBJ_ID_CACHE_REMOVED, (uintptr_t)node->avl.key, node->path);
+		free(node);
+	}
+
+	DEBUG("ID cache connected\n");
+
+	/* Add a few permanent IDs into the list. */
+	avl_object_add(ctx, 0, "(ubus core)");
+	avl_object_add(ctx, UBUS_SYSTEM_OBJECT_EVENT, "(ubus event)");
+	avl_object_add(ctx, UBUS_SYSTEM_OBJECT_ACL, "(ubus acl)");
+	avl_object_add(ctx, UBUS_SYSTEM_OBJECT_MONITOR, "(ubus monitor)");
+
+	ubus_register_event_handler(ubus, &ctx->ev, "ubus.object.*");
+	ctx->ev.cb = ubus_event_cb;
+	ubus_lookup(ubus, NULL, initial_object_list_cb, ctx);
+}
+
+void ubus_obj_id_cache_ubus_disconnected(struct ubus_obj_id_cache *ctx, struct ubus_context *ubus)
+{
+	struct ubus_object_id *node, *tmp;
+
+	/* ubus_unregister_event_handler() would normally be potentially appropriate
+	 * here, but it cannot (at the time of authorship) cope with ubus not
+	 * being connected at the time of invocation, nor can it survive a
+	 * reconnection cycle properly.
+	 */
+	memset(&ctx->ev, 0, sizeof(ctx->ev));
+
+	/* Reset cache */
+	avl_remove_all_elements(&ctx->avl, node, avl, tmp) {
+		if (ctx->cb)
+			ctx->cb(ctx, UBUS_OBJ_ID_CACHE_REMOVED, (uintptr_t)node->avl.key, node->path);
+		free(node);
+	}
+
+}
+
+void ubus_obj_id_cache_free(struct ubus_obj_id_cache *ctx)
+{
+	struct ubus_object_id *node, *tmp;
+
+	ubus_unregister_event_handler(ctx->ubus, &ctx->ev);
+	avl_remove_all_elements(&ctx->avl, node, avl, tmp) {
+		free(node);
+	}
+}
+
+const char *ubus_obj_id_cache_get(const struct ubus_obj_id_cache *server, uintptr_t id)
+{
+	struct ubus_object_id *node = avl_find_element(&server->avl, (void *)id, node, avl);
+	if (!node)
+		return NULL;
+
+	return node->path;
+}
+
diff --git a/jail/ubus_obj_id_cache.h b/jail/ubus_obj_id_cache.h
new file mode 100644
index 0000000..a5d2223
--- /dev/null
+++ b/jail/ubus_obj_id_cache.h
@@ -0,0 +1,41 @@
+/*
+ * Copyright (C) 2026 Tim Nordell <tnordell@airgain.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+#ifndef _JAIL_UBUS_OBJ_ID_CACHE_H
+#define _JAIL_UBUS_OBJ_ID_CACHE_H
+
+#include <libubox/avl.h>
+#include <libubus.h>
+
+enum UBUS_OBJ_ID_CACHE
+{
+  UBUS_OBJ_ID_CACHE_ADDED,
+  UBUS_OBJ_ID_CACHE_REMOVED,
+};
+
+struct ubus_obj_id_cache;
+typedef void (*ubus_obj_id_cache_cb)(struct ubus_obj_id_cache *, enum UBUS_OBJ_ID_CACHE, uint32_t id, const char *path);
+
+struct ubus_obj_id_cache {
+  struct ubus_context *ubus;
+	struct ubus_event_handler ev;
+	struct avl_tree avl;
+  ubus_obj_id_cache_cb cb;
+};
+
+void ubus_obj_id_cache_init(struct ubus_obj_id_cache *cache, ubus_obj_id_cache_cb cb);
+void ubus_obj_id_cache_ubus_connected(struct ubus_obj_id_cache *cache, struct ubus_context *ubus);
+void ubus_obj_id_cache_ubus_disconnected(struct ubus_obj_id_cache *cache, struct ubus_context *ubus);
+const char *ubus_obj_id_cache_get(const struct ubus_obj_id_cache *cache, uintptr_t id);
+void ubus_obj_id_cache_free(struct ubus_obj_id_cache *cache);
+
+#endif
-- 
2.53.0

